### DISCUSSION TITLE
Responsive design have should lower breakpoints by default

### DISCUSSION THREAD

**Discussion by Odafe88** [tailwindlabs/tailwindcss#11718] [Ideas]
**Responsive design have should lower breakpoints by default**

Most applications I have worked on have always had breakpoints as low as 320px. In my opinion I believe it helps with inclusivity for devices with smaller screen sizes. 

* **By default Tailwind CSS has its smallest breakpoint at 640px**: Having smaller breakpoints would save developers time and bring ease when it comes to responsive design building.
* **Also I don't think having multiple breakpoints after 1020px (I might be wrong) is totally necessary (I might be wrong) after all, it is the min-width**
* **Lastly shouldn't there be a way to group related responsive styles?**

**Comment by mybearworld** [comment]
Are there any actual devices still in use that have widths from 0px-320px?

**Reply by ImLunaHey** [reply]
window mode on desktop browsers comes to mind.

**Comment by hornetty6** [comment]
As i know including lower breakpoints, such as 320px, in your responsive design can indeed enhance inclusivity by ensuring that your application is usable on a wide range of devices, including those with smaller screens. [screen mirroring pubg mobile](https://www.screenmirroring.onl/how-to-play-pubg-on-smart-tv-using-screen-mirroring/) This is particularly important for ensuring a positive user experience on mobile devices, which are commonly used for web browsing.  

Tailwind CSS, like many other CSS frameworks, chooses default breakpoints based on what the framework's developers believe to be the most common screen sizes and device categories. However, these defaults can be adjusted to suit your project's needs. If you find that lower breakpoints are more suitable for your application, you can customize Tailwind CSS's breakpoints using the configuration file.

Grouping related responsive styles can indeed make your codebase more organized and maintainable. While Tailwind CSS offers utility classes for responsive design, you can also create custom utility classes that encapsulate related responsive styles. Additionally, you can use CSS preprocessors like Sass or Less to define mixins or functions for grouping responsive styles.

**Reply by mybearworld** [reply]
ChatGPT responses aren't helpful

**Comment by wolfspyre** [comment]
I'd agree that there really aught be a `default` breakpoint < 640... 

looking online today, I found the following data:
https://www.browserstack.com/guide/ideal-screen-sizes-for-responsive-design:
lists the following resolutions for mobile:
| Mobile resolution | Percentage |
|-------------------|------------|
| 360×800           | 11.01%     |
| 390×844           |  7.92%     |
| 414×896           |  5.55%     |
| 393×873           |  5.26%     |
| 412×915           |  5.00%     |

and the following for desktop:
|Desktop Resolution | Percentage|
|-------------------|-----------|
| 1920×1080         |  22.18%   |
| 1366×768          |  14.04%   | 
| 1440×900          |   6.41%   |
| 1280×720          |   5.45%   |
| 1280×1024         |   4.52%   |



https://yesviz.com/iphones.php: 
| device name    | viewport  | resolution |screen |pixel|px/i |css px/i |     CSS Media Query        |
|----------------|-----------|------------|-------|-----|-----|-----|--------------------------------|
|Apple iPhone 11 | 414 x 896 | 828 x 1792 | 6.1″ | 2.0 | 326 | 163 | Apple iPhone 11 |
|Apple iPhone 11 Pro|  375 x 812 | 1125 x 2436 | 5.8″ | 3.0 | 458 | 153 | Apple iPhone 11 Pro |
|Apple iPhone 11 Pro Max | 414 x 896 | 1242 x 2688 | 6.5″ | 3.0 | 458 | 153 | Apple iPhone 11 Pro Max |
|Apple iPhone 12 | 390 x 844 | 1170 x 2532 | 6.1″ | 3.0 | 460 | 153 | Apple iPhone 12 |
|Apple iPhone 12 mini | 360 x 780 | 1080 x 2340 | 5.4″ | 3.0 | 476 | 159 | Apple iPhone 12 Mini |
|Apple iPhone 12 Pro | 390 x 844 | 1170 x 2532 | 6.1″ | 3.0 | 460 | 153 | Apple iPhone 12 Pro |
|Apple iPhone 12 Pro Max | 428 x 926 | 1284 x 2778 | 6.7″ | 3.0 | 458 | 153 | Apple iPhone 12 Pro Max |
|Apple iPhone 13 | 390 x 844 | 1170 x 2532 | 6.1″ | 3.0 | 460 | 153 | Apple iPhone 13 |
|Apple iPhone 13 mini | 360 x 780 | 1080 x 2340 | 5.4″ | 3.0 | 476 | 159 | Apple iPhone 13 Mini |
|Apple iPhone 13 Pro | 390 x 844 | 1170 x 2532 | 6.1″ | 3.0 | 460 | 153 | Apple iPhone 13 Pro |
|Apple iPhone 13 Pro Max | 428 x 926 | 1284 x 2778 | 6.7″ | 3.0 | 458 | 153 | Apple iPhone 13 Pro Max |
|Apple iPhone 14 | 390 x 844 | 1170 x 2532 | 6.1″ | 3.0 | 460 | 153 | Apple iPhone 14 |
|Apple iPhone 14 Plus | 428 x 926 | 1284 x 2778 | 6.7″ | 3.0 | 458 | 153 | Apple iPhone 14 Plus |
|Apple iPhone 14 Pro | 393 x 852 | 1179 x 2556 | 6.1″ | 3.0 | 460 | 153 | Apple iPhone 14 Pro |
|Apple iPhone 14 Pro Max | 430 x 932 | 1290 x 2796 | 6.7″ | 3.0 | 460 | 153 | Apple iPhone 14 Pro Max |
|Apple iPhone 15 | 393 x 852 | 1179 x 2556 | 6.1″ | 3.0 | 461 | 153 | Apple iPhone 15 |
|Apple iPhone 15 Plus | 430 x 932 | 1290 x 2796 | 6.7″ | 3.0 | 460 | 153 | Apple iPhone 15 Plus |
|Apple iPhone 15 Pro | 393 x 852 | 1179 x 2556 | 6.1″ | 3.0 | 461 | 153 | Apple iPhone 15 Pro |
|Apple iPhone 15 Pro Max | 430 x 932 | 1290 x 2796 | 6.7″ | 3.0 | 460 | 153 | Apple iPhone 15 Pro Max |
|Apple iPhone 5 | 320 x 568 | 640 x 1136 | 4.0″ | 2.0 | 326 | 163 | Apple iPhone 5 |
|Apple iPhone 5c | 320 x 568 | 640 x 1136 | 4.0″ | 2.0 | 326 | 163 | Apple iPhone 5c |
|Apple iPhone 5s | 320 x 568 | 640 x 1136 | 4.0″ | 2.0 | 326 | 163 | Apple iPhone 5s |
|Apple iPhone 6 | 375 x 667 | 750 x 1334 | 4.7″ | 2.0 | 326 | 163 | Apple iPhone 6 |
|Apple iPhone 6s | 375 x 667 | 750 x 1334 | 4.7″ | 2.0 | 326 | 163 | Apple iPhone 6s |
|Apple iPhone 6s Plus | 414 x 736 | 1080 x 1920 | 5.5″ | 3.0 | 401 | 134 | Apple iPhone 6s Plus |
|Apple iPhone 7 | 375 x 667 | 750 x 1334 | 4.7″ | 2.0 | 326 | 163 | Apple iPhone 7 |
|Apple iPhone 7 Plus | 414 x 736 | 1080 x 1920 | 5.5″ | 3.0 | 401 | 134 | Apple iPhone 7 Plus |
|Apple iPhone 8 | 375 x 667 | 750 x 1334 | 4.7″ | 2.0 | 326 | 163 | Apple iPhone 8 |
|Apple iPhone 8 Plus | 414 x 736 | 1080 x 1920 | 5.5″ | 3.0 | 401 | 134 | Apple iPhone 8 Plus |
|Apple iPhone SE | 320 x 568 | 640 x 1136 | 4.0″ | 2.0 | 326 | 163 | Apple iPhone SE |
|Apple iPhone SE (2020) | 375 x 667 | 750 x 1334 | 4.7″ | 2.0 | 326 | 163 | Apple iPhone SE |
|Apple iPhone SE (2022) | 375 x 667 | 750 x 1334 | 4.7″ | 2.0 | 326 | 163 | Apple iPhone 14 SE 2022 |
|Apple iPhone X | 375 x 812 | 1125 x 2436 | 5.8″ | 3.0 | 458 | 153 | Apple iPhone X |
|Apple iPhone XR | 414 x 896 | 828 x 1792 | 6.1″ | 2.0 | 326 | 163 | Apple iPhone XR |
|Apple iPhone XS Max | 414 x 896 | 1242 x 2688 | 6.5″ | 3.0 | 458 | 153 | Apple iPhone XS Max |


this kinda adds some merit (from my perspective) of having a smaller breakpoint than 640 by default....  as pretty much every iphone's reported viewport is < 640w  

I'm sure those making these decisions did so with sound reasoninng.... and I'm missing something.... would someone mind helping me understand what it is I'm missing? 

I came to the conclusion that the guiding idea behind this might be 
> Anything with a width < 640 should be treated as mobile

... which could lead one to the perspective that there's not much need for further delineation.... 

which isn't **WRONG** ... but I suspect it has the added negative effect of subtly conveying the concept to test starting at 640x ... 

...(or maybe that's just how I, as an ignorant human, interpreted it)...

I suspect that the main upside of adding a smaller ie `xs` or `mbl` breakpoint would be that it helps people with initial planning... but ¯\\\_(ツ)_/¯  idunno, really.

**Reply by adamwathan** [reply]
Hey! So I did all the same research as you here which actually increased my confidence in choosing 640px as the lowest breakpoint.

You'll see in that table that aside from the iPhone 5 which was discontinued many years ago, all of the dimensions that appear are between 360px and 430px, which is only a 70px range.

Adding a breakpoint like say `xs` that targets 475px and below to me wouldn't be very useful, because there are effectively no devices in the 475px – 640px range anyways.

Of course people can resize their browser to any size, so just basing things off of common mobile viewport sizes doesn't take everything into account. At the time we made these decisions originally though we had to be more mindful of file size with Tailwind than we do now that everything is generated on demand.

Still I hesitate to introduce another smaller size because I think it makes people feel like they need to design for that size, and in my experience you just kinda don't.

What we usually do if anything is just add a max-width of say 475px or so to the mobile design, so that at 600px wide you just see a narrow design with some extra breathing room on the sides. I've found it's rare to even be able to do much with that extra space — it tends to not be enough to introduce another column of content until you get up to that 640px range anyways.

Anyways — I wouldn't say it's something we're totally opposed to ever adding but it's never really been obvious to me what the number should be or why.

> but I suspect it has the added negative effect of subtly conveying the concept to test starting at 640x ...

This is definitely not the intention — the idea with the Tailwind defaults is that you build things _mobile-first_, so a class with no breakpoint is targeting mobile, and then you add `sm:` when you need to make tweaks for screens 640px and up. Not specifying a breakpoint at all _is_ the mobile breakpoint in Tailwind:

https://tailwindcss.com/docs/responsive-design#working-mobile-first

**Reply by wolfspyre** [reply]
Hi adam!

firstly; THANKS for sharing your thoughts here, i imagine that took you a nontrivial amount of time to write, and i appreciate the investment.


re: max-w 475 that’d be a good starting-point guide for peeps i suspect 

speaking from my initial meander into mobile ui layout
my internal narrative after surfing through the docs a few times was
> “okay anything under 640 is mobile… so what do I use as a mobile mockup dimensions? “

…And I reached for 600w 
(which really wasn’t a great strategy… I saw no mention of anything else whilst reading, and admittedly did very little extra research initially)

Was my strategy lazy and misinformed?

ABSOLUTELY… but i didn’t know what to pick at the time, and used 640 as an inferred min-size-of-non-mobile-viewport

 agree it’s not the intention… 

however in the absence of a guide, someone not acutely familiar with design may not pick up on the hint.

i think if the docs offered the max-w guidance mentioned above,  that mighta steered me in a better initial trajectory, but i really dunno …

regardless… thanks for sharing your perspective!

W

**Comment by tylerrrkd** [comment]
my config will help you.

```
const defaultTheme = require('tailwindcss/defaultTheme');

/** @type {import('tailwindcss').Config} */
module.exports = {
  content: ['./src/**/*.{js,ts,jsx,tsx,mdx}'],
  theme: {
    extend: {
      screens: {
        // => @media (min-width: 0px and max-width: 639px) { ... }
        xs: { min: '0px', max: '639px' },
        /**
         * @see https://tailwindcss.com/docs/screens#adding-smaller-breakpoints
         */
        ...defaultTheme.screens,
      },
    },
  },
  plugins: [],
};
```

**Reply by wolfspyre** [reply]
the point isn’t just to add a smaller breakpoint only in my config, (which i’ve already done, somewhat similarly, so seeing yours does validate my thought… thanks for sharing!)

but rather to bring the conversation up to try and understand what i’m missing that this isn’t already a thing 
 :)

**Reply by lucywho** [reply]
This is really helpful, thanks

**Reply by daliborfilus** [reply]
Your `xs` breakpoint is the same as the "no prefix" breakpoint, and thus completely useless. People, don't add this to your config, it makes no sense.
